### PR TITLE
fix: client side window

### PR DESCRIPTION
**Pull request by Meemaw**
## What kind of change does this PR introduce?

Bug fix

## What is the current behaviour?

Using libraries that rely on `window` object, e.g. `react-hook-form` (https://github.com/react-hook-form/react-hook-form/blob/398bb117b6dac0e9d4c5bc2d7c2ed328c932332e/src/useForm.ts#L77) won't work

## What is the new behaviour?

Using libraries that rely on `window` object works.

## Does this PR introduce a breaking change?

No

## Please check if the PR fulfills these requirements:

- [x] Tests for the changes have been added
- [x] Docs have been added / updated

### CODE REVIEW

File: src/loadPage.ts [hunk 1]
```diff
@@ -19,13 +13,23 @@ export function loadPage({
   // @ts-ignore
   delete global.window;
 
-  // @NOTE Here we have to remove pagePath's trailing "/"
-  const page = require(path.resolve(pagesDirectory, pagePath.substring(1)));
+  require('next/dist/next-server/lib/side-effect');
```
Comment (Meemaw) [inline src/loadPage.ts:16]: Before requiring the whole page, we require only the module that we want to trick to think is on server.

**Comment by toomuchdesign** [comment]
Hi @Meemaw,
how often do we need `loadPage` to fake a non-browser environment?

Could adding a `nonBrowserEnvironment` (sorry for the bad naming) work, to temporarily make `global.window` unavailable when we load certain files (like custom `_document`)?

**Comment by Meemaw** [comment]
@toomuchdesign  I'm not sure if I understand what you mean.

We need to fake non-browser environment for NextJS to properly render page when `useDocument=true` (head tags).

Previously we did that before loading the whole page, but that was problematic, because it meant window was undefined also for all the modules and external libraries that page is using.

New behaviour is to isolate that only to apply to the specific NextJS module `next/dist/next-server/lib/side-effect`.

**Comment by toomuchdesign** [comment]
I see, thanks @Meemaw!
I checked out the PR and it seems that `trickNextJsIsServer` is not particularly related to `getPage`. It's just Next which needs `next/dist/next-server/lib/side-effect` to be evaluated in non-browser mode before the document is rendered.

What if we moved `trickNextJsIsServer` somewhere into `renderDocument` and maybe rename it to something like `evaluateSideEffect` or whatever that file does?

As bonus we would perform this nasty trick only when `useDocument` set to true and leave the other use-case untouched.

**Comment by Meemaw** [comment]
Agree, it is not related to `getPage`. However, we cannot move it to some step after this because after the `loadPage` is required in the next step it is already too late. At that point all modules are imported and we cannot influence their top level expressions anymore.

We can however add a check to call this only if `useDocument=true` and add a comment why we call it at this point.

**Comment by toomuchdesign** [comment]
If it's needed to be loaded before page is fetched we might:
- assume it's `loadPage` responsibility (and therefore add a flag which we set to `true` when `useDocument` is `true`)
- assume it should not go in `loadPage` and trigger the require trick [somewhere here](https://github.com/toomuchdesign/next-page-tester/blob/fix-client-side-window/src/getPageObject.ts#L14) in `getPageObject` object

How do you see it?

**Comment by Meemaw** [comment]
I did come changes in https://github.com/toomuchdesign/next-page-tester/pull/44/commits/98939a777a5b9517d47ecbfe77892b5691e0083f so we are conditionally loading it. I still think it is actually a good place to execute it as it is related to module loading/resolving.

**Review by toomuchdesign — APPROVED** [review]
Bien sûr, looks good to me :)

**Pull request merged**